### PR TITLE
Refactor visualization functions

### PR DESCRIPTION
**Pull request by undivisible**
Combines `generate_topics_visualization` into `generate_visualization` in `backend/scripts/rag/current.py` as requested.
- Implements `memories: List[Conversation] | None = None` support to allow calling either `get_data2` or `get_data`.
- Fixes edge cases with `len(topics) == 0` preventing slicing exceptions.
- Removes redundant code.
- Updates entrypoint in `__main__` to run `generate_visualization([])`.

---
*PR created automatically by Jules for task [12965376793190174944](https://jules.google.com/task/12965376793190174944) started by @undivisible*

<a href="https://cubic.dev/pr/BasedHardware/omi/pull/11341?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

---

> [!NOTE]
> **Low Risk**
> Changes are limited to a RAG script helper and new unit tests; no production API or auth paths.
> 
> **Overview**
> **Refactors RAG embedding HTML visualization** in `backend/scripts/rag/current.py` by removing `generate_topics_visualization` and folding its behavior into `generate_visualization`.
> 
> `generate_visualization` now takes optional `memories`; when set it uses `get_data2`, otherwise `get_data`. Topic query embeddings are only stacked when `topics` is non-empty, fixing slice errors when `topics` is `[]`. The function **no-ops** when there are no embeddings, invalid embedding shape, or fewer than three UMAP samples, and tunes UMAP (`init`, `n_neighbors`) for small point counts. The `__main__` entrypoint calls `generate_visualization([])`.
> 
> Adds **`backend/tests/unit/test_rag_visualization.py`** with isolated module loads and tests for topic vs no-topic paths, empty/insufficient data skips, and the `memories` → `get_data2` branch.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 286335137b17a5c17080516cc6caf6933d9573aa. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

Failure-Class: none

### CODE REVIEW

**Review by undivisible — COMMENTED** [review]
The empty-topic entrypoint can load all stored vectors and then run UMAP on empty or too-small data, which can crash on a fresh or sparse store. Please guard empty/insufficient embeddings before UMAP and add a regression test for generate_visualization([]).

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for iterating on this. I reviewed the current head and there is one blocking cleanup item before this should merge.

Blocking:
- `rewrite_commit.py` should not be committed. It is a repository-root helper that writes directly to `.git/COMMIT_EDITMSG` at import/runtime, is not part of the RAG visualization implementation, and appears to have been used only to try to rewrite commit metadata. Please remove this file and fix the commit metadata through the normal PR/commit workflow instead. This also aligns with the current Hygiene failure, which is reporting that the `fix:` commits still do not have a valid `Failure-Class: FC-<slug> | new | none` declaration in the actual commit history.

Other observations:
- `backend/scripts/rag/current.py` is moving in the right direction: consolidating `generate_topics_visualization` into `generate_visualization`, allowing `memories=None` to use `get_data`, skipping topic embedding/slicing when `topics` is empty, and bounding UMAP `n_neighbors` address the sparse/empty visualization path. I did not see an obvious production regression in that code path from static review.
- `backend/tests/unit/test_rag_visualization.py` adds coverage for the new pre-UMAP guards, but it is an AST/source-shape test rather than an executable behavior test. That is acceptable as a lightweight guard here, but a future follow-up could make this stronger by stubbing `get_data`/`umap.UMAP` and calling `generate_visualization([])` directly.

Leaving this as changes requested for the stray root-level commit-rewrite helper and the still-failing Hygiene/failure-class gate.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for the follow-up. The visualization refactor itself looks reasonable, but there is still one repository-hygiene blocker before this should merge.

Blocking:
- `rewrite_commit.py` is still present at the repository root. It executes `rewrite()` at import/runtime and writes directly to `.git/COMMIT_EDITMSG`, which is unrelated to the RAG visualization code and should not be committed. Please remove this helper and handle commit metadata through the normal PR/commit workflow.

File-by-file notes:
- `backend/scripts/rag/current.py`: consolidating `generate_topics_visualization` into `generate_visualization`, making `memories` optional, skipping topic embeddings when `topics` is empty, and bounding `n_neighbors` for small embedding sets all look aligned with the intended refactor. I did not see a production-path regression in this script from static review.
- `backend/tests/unit/test_rag_visualization.py`: the new isolated tests exercise both topic-augmented and no-topic visualization paths by stubbing `get_data`, `umap.UMAP`, and Plotly, which is useful coverage for the edge cases changed here.
- `rewrite_commit.py`: this remains the blocking issue because it is a root-level helper that mutates Git commit-message state and is not part of the backend/RAG visualization implementation.

Leaving this as changes requested for the stray commit-message rewrite helper and the related workflow/hygiene failure. Once that file is removed and the checks are green, this looks like a focused backend refactor.

---
*by AI on behalf of David — leaving this for maintainer review once the workflow/hygiene blocker is fixed.*

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for continuing to tighten this up. The backend visualization refactor itself looks focused, but the current head still has one blocking repository-hygiene/workflow artifact that should not merge.

Blocking:
- `rewrite_commit.py` is still committed at the repository root. It imports at top level and calls `rewrite()` immediately, writing directly to `.git/COMMIT_EDITMSG`. That is unrelated to the RAG visualization path and should be removed from the PR rather than shipped as source.

File-by-file notes:
- `backend/scripts/rag/current.py`: consolidating the old topic-only visualization path into `generate_visualization`, using `get_data()` when `memories` is omitted, skipping topic embeddings/slicing for `topics=[]`, and bounding UMAP `n_neighbors` all look aligned with the intended sparse/empty visualization behavior.
- `backend/tests/unit/test_rag_visualization.py`: the isolated tests now exercise both a topic-augmented small memory set and the topic-free path by stubbing `get_data`, UMAP, and Plotly, which is useful behavioral coverage for the changed slicing/UMAP boundaries.
- `rewrite_commit.py`: this remains the blocker because it is a root-level commit-message rewrite helper with side effects against `.git/COMMIT_EDITMSG`, not backend visualization code.

Leaving this as changes requested for the stray workflow helper. Once `rewrite_commit.py` is removed, the remaining backend refactor/test changes look reasonable from this review.

---
*by AI on behalf of David — leaving this for human maintainer review only if there is a workflow/source-hygiene question after the helper is removed.*

**Comment by Git-on-my-level** [comment]
Thanks for cleaning this up — the previous repository-hygiene blocker appears resolved on this head.

Current review notes:
- `backend/scripts/rag/current.py`: the refactor now folds the old topic-only visualization path into `generate_visualization`, keeps the app path working via `memories: List[Conversation] | None`, uses `get_data()` for the no-memory entrypoint, avoids topic embedding/slicing when `topics=[]`, and returns before UMAP for empty/non-2D/too-small embedding sets. The `n_neighbors` cap and `init='random'` for the 3-point case are appropriate guards for sparse visualization data.
- `backend/tests/unit/test_rag_visualization.py`: the new isolated tests cover the topic-augmented path, the no-topic path preserving all memory points, empty-data early return, and the `memories`/`get_data2` branch. I ran the focused test from the repository root with `PYTHONPATH=backend`; it passed (`4 passed`). Running the same file from inside `backend/` fails because the test passes `backend/scripts/rag/current.py` to `load_module_fresh`, which resolves to `backend/backend/scripts/rag/current.py` from that working directory. That looks like command/working-directory sensitivity rather than a production regression, and the repo-root invocation matches the hermetic checks here.

The earlier `rewrite_commit.py` / commit-message helper concern is resolved: the current diff only contains `backend/scripts/rag/current.py` and `backend/tests/unit/test_rag_visualization.py`. I’m clearing the stale workflow blocker and leaving a positive signal for the focused backend refactor/tests rather than formal approval.

---
*by AI on behalf of David — leaving final merge judgment to the maintainer if they want the backend test to be runnable from additional working directories.*

File: backend/scripts/rag/current.py [hunk 1]
```diff
@@ -157,25 +102,52 @@ def get_data2(topics: List[str], retrieved_memories: List[Conversation]) -> dict
...
-
-    umap_transform = cast(Any, umap.UMAP(n_components=2, random_state=0, transform_seed=0))
+    if memories is not None:
+        data = get_data2(topics, memories)
+    else:
+        data = get_data(topics)
+
+    embedding_values = [item[1] for item in data.values()]
+    if not embedding_values:
+        return
```
[the remark this answers is not in the document]
Comment (undivisible) [reply backend/scripts/rag/current.py:127]: Deferred intentionally: this is a local analysis script (backend/scripts/rag/current.py) run manually, not a production request path. The empty/insufficient-sample guards are deliberate no-op exits for the developer running the tool, and `record_fallback` is a production telemetry contract for provider/mode failovers. Adding it here would emit metrics/log noise with no operational consumer. If visualization generation ever moves behind a served endpoint, the guard should be instrumented there.

File: backend/tests/unit/test_rag_visualization.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    module.get_query_marker = MagicMock(return_value=object())
+    module.generate_html_visualization = MagicMock()
+    module.go = SimpleNamespace(Scatter=MagicMock())
+    figure = MagicMock()
+    module.make_subplots = MagicMock(return_value=figure)
+    umap_instance = MagicMock()
+    umap_instance.fit_transform.return_value = np.array([[0.0, 0.0], [1.0, 1.0], [2.0, 2.0]])
+    umap_module.UMAP.return_value = umap_instance
+
+    module.generate_visualization(['topic'])
```
[the remark this answers is not in the document]
Comment (undivisible) [reply backend/tests/unit/test_rag_visualization.py:115]: Covered on the current head: test_generate_visualization_skips_below_three_samples supplies one embedding and asserts UMAP and HTML are not called.

File: backend/scripts/rag/current.py [hunk 3]
```diff
@@ -156,26 +103,34 @@ def get_data2(topics: List[str], retrieved_memories: List[Conversation]) -> dict
...
+    if memories is not None:
+        data = get_data2(topics, memories)
+    else:
+        data = get_data(topics)
+
     all_embeddings = cast(Any, np.array([item[1] for item in data.values()]))
 
-    topic_embeddings = [openai_embeddings.embed_query(topic) for topic in topics]
-    all_embeddings = cast(Any, np.vstack([all_embeddings] + topic_embeddings))
+    if topics:
```
[the remark this answers is not in the document]
Comment (undivisible) [reply backend/scripts/rag/current.py]: The current regression suite includes the empty-topic path and verifies both the fitted input and the marker input, so this duplicate coverage request is satisfied.

File: backend/scripts/rag/current.py [hunk 4]
```diff
@@ -156,26 +103,34 @@ def get_data2(topics: List[str], retrieved_memories: List[Conversation]) -> dict
...
+        topic_embeddings = [openai_embeddings.embed_query(topic) for topic in topics]
+        all_embeddings = cast(Any, np.vstack([all_embeddings] + topic_embeddings))
```
[the remark this answers is not in the document]
Comment (undivisible) [reply backend/scripts/rag/current.py]: Covered on the current head: test_generate_visualization_keeps_all_memory_points_without_topics exercises topics=[] with three memories, asserts all three points reach get_markers, and verifies HTML generation.

File: backend/tests/unit/test_rag_visualization.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    module.get_markers = MagicMock(return_value=object())
+    module.generate_html_visualization = MagicMock()
+    figure = MagicMock()
+    module.make_subplots = MagicMock(return_value=figure)
+    umap_instance = MagicMock()
+    umap_instance.fit_transform.return_value = np.array([[0.0, 0.0], [1.0, 1.0], [2.0, 2.0]])
+    umap_module.UMAP.return_value = umap_instance
+
+    module.generate_visualization([])
+
```
[the remark this answers is not in the document]
Comment (undivisible) [reply backend/tests/unit/test_rag_visualization.py]: Covered on the current head: the empty-topic regression asserts the exact marker array passed to get_markers, so a zero-length slice would fail.

File: backend/scripts/rag/current.py [hunk 6]
```diff
@@ -157,25 +102,51 @@ def get_data2(topics: List[str], retrieved_memories: List[Conversation]) -> dict
...
-    all_embeddings = cast(Any, np.array([item[1] for item in data.values()]))
-
-    topic_embeddings = [openai_embeddings.embed_query(topic) for topic in topics]
-    all_embeddings = cast(Any, np.vstack([all_embeddings] + topic_embeddings))
-
-    umap_transform = cast(Any, umap.UMAP(n_components=2, random_state=0, transform_seed=0))
+    if memories is not None:
+        data = get_data2(topics, memories)
+    else:
+        data = get_data(topics)
```
[the remark this answers is not in the document]
Comment (undivisible) [reply backend/scripts/rag/current.py]: Covered on the current head: test_generate_visualization_uses_retrieved_memories patches and asserts the production get_data2 seam.
